libs/common/RespReadUtils.cs

```diff
@@ -4,156 +4,363 @@
+        private static bool TryReadSign(byte* ptr, out bool negative)
         {
-            len = -1;
-            if (ptr + 3 >= end)
-                return false;
+            negative = (*ptr == '-');
+            return negative || (*ptr == '+');
+        }
 
-            Debug.Assert(*ptr == '$');
-            ptr++;
-            bool neg = *ptr == '-';
-            int ksize = *ptr++ - '0';
-            while (*ptr != '\r')
+        /// <summary>
+        /// Tries to read an unsigned 64-bit integer from a given ASCII-encoded input stream.
+        /// </summary>
+        /// <param name="ptr">Pointer to the beginning of the ASCII encoded input string.</param>
+        /// <param name="end">The end of the string to parse.</param>
+        /// <param name="value">If parsing was successful, contains the parsed ulong value.</param>
+        /// <param name="bytesRead">If parsing was successful, contains the number of bytes that were parsed.</param>
+        /// <returns>
+        /// True if a ulong was successfully parsed, false if the input string did not start with
+        /// a valid integer or the end of the string was reached before finishing parsing.
+        /// </returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool TryReadUlong(ref byte* ptr, byte* end, out ulong value, out ulong bytesRead)
+        {
+            bytesRead = 0;
+            value = 0;
+            var readHead = ptr;
+
+            // Fast path for the first 19 digits.
+            // NOTE: UINT64 overflows can only happen on digit 20 or later (if integer contains leading zeros).
+            var fastPathEnd = ptr + 19;
+            while (readHead < fastPathEnd)
             {
-                Debug.Assert(*ptr >= '0' && *ptr <= '9');
-                ksize = ksize * 10 + *ptr++ - '0';
-                if (ptr >= end)
+                if (readHead > end)
+                {
                     return false;
+                }
+
+                var nextDigit = (uint)(*readHead - '0');
+                if (nextDigit > 9)
+                {
+                    goto Done;
+                }
+
+                value = (10 * value) + nextDigit;
+
+                readHead++;
             }
 
-            ptr += 2;
-            if (ptr > end)
-                return false;
-            Debug.Assert(*(ptr - 2) == '\r');
-            Debug.Assert(*(ptr - 1) == '\n');
-            len = neg ? -ksize : ksize;
+            // Parse remaining digits, while checking for overflows.
+            while (true)
+            {
+                if (readHead > end)
+                {
+                    return false;
+                }
+
+                var nextDigit = (uint)(*readHead - '0');
+                if (nextDigit > 9)
+                {
+                    goto Done;
+                }
+
+                if ((value == 1844674407370955161UL && ((int)nextDigit > 5)) || (value > 1844674407370955161UL))
```

Comment:
untested, but I wonder if `value = checked((10 * value) + nextDigit);` might be clearer and more consistent here (using CPU overflow flag, and well-known `OverflowException`)

Reply:
I agree that it would be nice to use `checked` for this. However, the problem that I saw with using `checked` is that we need to convert the exception into a `RespParsingException`, so that we know to forward the error to the client. But catching and rethrowing is prohibitively expensive. The check above should be relatively close in performance to using `checked` for valid 20-digit ulong values.

Reply:
Why convert to `RespParsingException`, if we can instead catch that specific exception in the outer block?

Reply:
The connection will be closed for any exception regardless of its type. However, the type of exception determines how we are handling the response to the client. In the parsing case we would like to output as a protocol error to the user - but if an overflow happens anywhere outside of parsing this should be logged as a critical error on the server side only.

Of course, we could technically write logic to match the `OverflowException` back to the parsing functions - but that's a bad design for many many reasons. We would also lose important contextual information to output meaningful error messages. E.g., this is what this PR will throw:

`-ERR Protocol Error: Unable to parse integer. The given number is larger than allowed: 3000000000000000000`

Reply:
As for readability, maybe it would be useful to add a comment to remind the reader that UINT64 Max is 1844674407370955161**5**. Any opinions on this?

Reply:
We don't use 'checked' anywhere else in the codebase, and can keep it that way. This we could catch overflow exception and kill the connection. All assuming that is indeed more efficient ofc.

Reply:
I would think overflow anywhere in the call path should kill the connection anyway. I can't think of a place where overflow could occur and we would want to continue processing.

Reply:
`OverflowException` can be thrown in other contexts as well. So, even if we don't use `checked` anywhere else, that does not guarantee an `OverflowException` caught in `TryConsumeMessages()` is a parsing exception that should be forwarded to the client.

Personally, I am not sure it's worth giving up clean error handling for the gains we are getting from using checked (keep in mind this is also just in isolation).  See below for a quick benchmark run.

<details>
<summary>TryReadLong: checked vs RespParsingException</summary><p>

| Method              | Job    | EnvironmentVariables | Runtime  | testCase               | Mean      | Error     | StdDev    | Median    |
|-------------------- |------- |--------------------- |--------- |----------------------- |----------:|----------:|----------:|----------:|
| TryReadUlong        | .NET 6 | Empty                | .NET 6.0 | 0\r\n                  |  2.101 ns | 0.0026 ns | 0.0024 ns |  2.101 ns |
| TryReadUlongChecked | .NET 6 | Empty                | .NET 6.0 | 0\r\n                  |  1.656 ns | 0.0034 ns | 0.0030 ns |  1.655 ns |
| TryReadUlong        | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 | 0\r\n                  |  2.124 ns | 0.0016 ns | 0.0014 ns |  2.123 ns |
| TryReadUlongChecked | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 | 0\r\n                  |  1.955 ns | 0.0030 ns | 0.0024 ns |  1.954 ns |
| TryReadUlong        | .NET 6 | Empty                | .NET 6.0 | 18446(...)615\r\n [22] | 13.097 ns | 0.0076 ns | 0.0067 ns | 13.096 ns |
| TryReadUlongChecked | .NET 6 | Empty                | .NET 6.0 | 18446(...)615\r\n [22] | 12.247 ns | 0.0059 ns | 0.0052 ns | 12.247 ns |
| TryReadUlong        | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 | 18446(...)615\r\n [22] | 12.823 ns | 0.0064 ns | 0.0060 ns | 12.825 ns |
| TryReadUlongChecked | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 | 18446(...)615\r\n [22] | 12.145 ns | 0.0096 ns | 0.0090 ns | 12.144 ns |
| TryReadUlong        | .NET 6 | Empty                | .NET 6.0 | 2147483647\r\n         |  7.475 ns | 0.0448 ns | 0.0419 ns |  7.473 ns |
| TryReadUlongChecked | .NET 6 | Empty                | .NET 6.0 | 2147483647\r\n         |  7.444 ns | 0.0703 ns | 0.0658 ns |  7.427 ns |
| TryReadUlong        | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 | 2147483647\r\n         |  7.223 ns | 0.0499 ns | 0.0442 ns |  7.249 ns |
| TryReadUlongChecked | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 | 2147483647\r\n         |  6.320 ns | 0.1547 ns | 0.3089 ns |  6.167 ns |
</p>
</details>

Reply:
Which other context? It only gets thrown in `checked` and we don't use that anywhere else.

Reply:
What I meant is that it can be thrown by functions that use `checked` internally. For example, `Convert.ToInt64()` can throw an `OverflowException` as well.

Reply:
sounds good, resolving this conversation.